totalRP3_Extended/aura/aura.lua

```diff
@@ -0,0 +1,1009 @@
+			end
+		else -- "="
+			newExpiry = now + duration;
+		end
+		if newExpiry ~= currentExpiry then
+			aura.persistent.expiry = newExpiry;
+			if newExpiry > now then
+				aura.expired = nil;
+			end
+			self:Update();
+		end
+	end,
+	
+	SetAuraVariable = function(self, aura, opType, varName, value)
+		local shouldUpdate = false;
+		if opType == "[=]" then
+			if not aura.persistent.vars or aura.persistent.vars[varName] == nil then
+				aura.persistent.vars = aura.persistent.vars or {};
+				aura.persistent.vars[varName] = value;
+				shouldUpdate = aura.hasDynamicOverlay;
+			end
+		elseif opType == "=" then
+			aura.persistent.vars = aura.persistent.vars or {};
+			aura.persistent.vars[varName] = value;
+			shouldUpdate = aura.hasDynamicOverlay;
+		else
+			aura.persistent.vars = aura.persistent.vars or {};
+			local op1 = tonumber(aura.persistent.vars[varName]) or 0;
+			local op2 = tonumber(value) or 0;
+			if opType == "+" then
+				aura.persistent.vars[varName] = op1 + op2;
+			elseif opType == "-" then
+				aura.persistent.vars[varName] = op1 - op2;
+			elseif opType == "x" then
+				aura.persistent.vars[varName] = op1 * op2;
+			elseif opType == "/" then
+				aura.persistent.vars[varName] = op1 / op2;
+			end
+			shouldUpdate = aura.hasDynamicOverlay;
+		end
+		
+		if shouldUpdate then
+			self:Update();
+		end
+	end,
+	
+	GetAuraColorFromClass = function(self, class)
+		if class.BA.CO then
+			local colorCached = {
+				h = class.BA.CO
+			}
+			colorCached.r, colorCached.g, colorCached.b = hexaToFloat(class.BA.CO);
+			return colorCached;
+		else
+			return nil;
+		end
+	end,
+	
+	InsertNewAura = function(self, auraId, class)
+		local now = self:Now();
+		local newAuraPersistent = {
+			id = auraId,
+			expiry = now + math.abs(class.BA.DU or math.huge),
+			lastTick = now,
+			dormantSince = now
+		};
+		tinsert(self.currentProfile.auras, newAuraPersistent);
+		local newAura = {
+			class = class,
+			persistent = newAuraPersistent,
+			color = self:GetAuraColorFromClass(class),
+			initialize = true
+		};
+		self:AnalyzeAuraClass(newAura);
+		tinsert(self.activeAuras, newAura);
+		self:RegisterAuraEvents(newAura);
+		self:Update();
+	end,
+	
+	CancelAura = function(self, auraId)
+		local aura = self:FindAura(auraId);
+		if aura then
+			aura.cancel = true;
+			self:Update();
+			return true;
+		end
+		return false;
+	end,
+	
+	RemoveAura = function(self, auraId)
+		local aura = self:FindAura(auraId);
+		if aura then
+			aura.persistent.invalid = true;
+			self:UnregisterAuraEvents(aura);
+			self:Update();
+			return true;
+		end
+		return false;
+	end,
+	
+	FindAura = function(self, auraId)
+		for i=1,#self.activeAuras do
+			if not self.activeAuras[i].persistent.invalid and self.activeAuras[i].persistent.id == auraId then
+				return self.activeAuras[i];
+			end
+		end
+		return nil;
+	end,
+	
+	GetAuraAt = function(self, index)
+		local j = 1;
+		for i = 1,#self.activeAuras do
+			if not self.activeAuras[i].persistent.invalid then
```

Comment:
This pattern of iterating over active auras that aren't invalid seems quite common. Consider making this a custom iterator.

```lua
local function NextValidAura(auras, index)
    index = index + 1;

    while auras[index] do
        local aura = auras[index];

        if not aura.persistent.invalid then
            return index, aura;
        else
            index = index + 1;
        end
    end
end

local function EnumerateValidAuras(auras)
    return NextValidAura, auras, 0;
end

-- Loops can then become this:

for index, aura in EnumerateValidAuras(self.activeAuras) do
    print(index, aura)
end
```